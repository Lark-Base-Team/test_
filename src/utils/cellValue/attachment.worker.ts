async function downloadFile(url: string) {
  return new Promise<File>((resolve, reject) => {
    const xhr = new XMLHttpRequest()
    xhr.open("GET", url, true)
    xhr.responseType = "blob"
    xhr.onload = function () {
      if (this.status === 200) {
        resolve(this.response)
      } else {
        reject(this.response)
      }
    }
    xhr.send()
  })
}
