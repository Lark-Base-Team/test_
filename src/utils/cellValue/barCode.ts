

export function barCode(value: string) {
  return {
    text: String(value),
    type: "text",
  }
};