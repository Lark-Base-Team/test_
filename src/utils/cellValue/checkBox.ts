import type { IOpenCheckbox } from "@lark-base-open/js-sdk";

export const defaultBoolValue = {
  true: ["是", "True", "true", "TRUE", "1", "☑️"],
  false: ["否", "False", "false", "FALSE", "0", ""],
}

export function checkBox(value: string, boolValue: { true: string[]; false: string[]; } = defaultBoolValue): IOpenCheckbox {
  if (!value) return false;
  // const boolValue = config?.boolValue;
  if (boolValue.true.includes(value)) return true;
  if (boolValue.false.includes(value)) return false;
  return false
}