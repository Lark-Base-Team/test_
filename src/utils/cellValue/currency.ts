import { checkers } from "@lark-base-open/js-sdk"

export function currency(value: string): number | null {
  const res = Number(value.match(/-?\d+\.?\d*/g))
  if (checkers.isNumber(res)) return res;
  return null
}