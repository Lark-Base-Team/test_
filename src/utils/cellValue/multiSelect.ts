import {
  IMultiSelectFieldMeta,
  IOpenMultiSelect,
  IWidgetTable,
  checkers
} from "@lark-base-open/js-sdk";

export const defaultSeparator = ","

export function multiSelect(
  value: string,
  field: IMultiSelectFieldMeta,
  // table: IWidgetTable,
  separator: string = ","
): IOpenMultiSelect {
  console.log("separator", separator)
  if(value === "") return []
  console.log("multiSelect", value)
  const selection = Array.from(value.split(separator))
  const res: IOpenMultiSelect = [];
  // let fieldItem = null;
  // let fieldItem = await table.getFieldMetaById(field.id) as IMultiSelectFieldMeta;
  for (const v of selection) {
    let id = field.property.options.find((option) => option.name === v)?.id ?? "";
    // if(id === "") {
    //   if(!fieldItem) fieldItem = await table.getFieldMetaById(field.id) as IMultiSelectFieldMeta;
    //   id = fieldItem.property.options.find((option) => option.name === v)?.id ?? "";
    // }
    res.push({
      text: v,
      id,
    });
  }
  if (checkers.isMultiSelect(res)) return res;
  return [];
}