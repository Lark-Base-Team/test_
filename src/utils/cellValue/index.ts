import {
  IOpenCellValue,
  FieldType,
  IMultiSelectFieldMeta,
  ISingleSelectFieldMeta,
} from "@lark-base-open/js-sdk";
import { fieldMap } from "@/types/types";
import { dateTime, dateDefaultFormat } from "./date";
import { multiSelect } from "./multiSelect";
import { singleSelect } from "./singleSelect";
import { checkBox, defaultBoolValue } from "./checkBox";
import { text } from "./text";
import { url } from "./url";
import { phone } from "./phone";
import { currency } from "./currency";
import { progress } from "./progress";
import { rating } from "./rating";
import { barCode } from "./barCode";
import {User} from "./user";

export function getCellValue(
  fieldMap: fieldMap,
  value: string,
): IOpenCellValue {
  const type = fieldMap.field.type;
  const field = fieldMap.field;
  const config = fieldMap.config;
  if (!value) return value;
  switch (type) {
    case FieldType.Url:
      return url(value);
    case FieldType.DateTime:
      return dateTime(value, config?.format ?? dateDefaultFormat);
    case FieldType.SingleSelect:
      return singleSelect(value, field as ISingleSelectFieldMeta);
    case FieldType.MultiSelect:
      return multiSelect(
        value,
        field as IMultiSelectFieldMeta,
        config?.separator ?? ","
      );
    case FieldType.Number:
      return Number(value.match(/-?\d+\.?\d*/g));
    case FieldType.Currency:
      return currency(value);
    case FieldType.Progress:
      return progress(value);
    case FieldType.Rating:
      return rating(value);
    case FieldType.Checkbox:
      return checkBox(value, config?.boolValue ?? defaultBoolValue);
    case FieldType.Text:
      return text(value) as IOpenCellValue;
    case FieldType.Barcode:
      return barCode(value) as IOpenCellValue;
    case FieldType.Phone:
      return phone(value);
    case FieldType.User:
      return User(value);
    default:
      return value;
  }
}
