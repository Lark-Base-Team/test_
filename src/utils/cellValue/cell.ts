import {
  FieldType,
  type IField,
  type ICell,
  type IOpenCellValue,
} from "@lark-base-open/js-sdk"
import type { fieldMap } from "@/types/types"
import { Error, FieldNameList } from "@/utils"

interface Progress {
  total: number
  loaded: number
  message?: string
}

export interface AsyncParams<T extends any = any> {
  data: Array<T>
  onProgress?: (progress: Progress) => void
  onError?: (error: any) => void
}

export enum FieldRole {
  AUTO = "auto",
  HAS_OPTIONS = "hasOptions",
  ASYNC = "async",
  NORMAL = "normal",
}

interface TranslatorOptions<
  T extends IField = any,
  K extends ICell = any,
  N = any,
> {
  fieldType: FieldType
  translate?: (
    value: string,
    field: T,
    config?: fieldMap["config"],
  ) => Promise<K | null>
  refresh?: () => void
  normalization: (value: string, config?: fieldMap["config"]) => Promise<N>
  name: string
  cache?: Record<string, any> | Array<any>
  roles?: Array<FieldRole>
  asyncMethod?: (
    options: AsyncParams,
    config?: fieldMap["config"],
  ) => Promise<N>
}

interface Translator extends TranslatorOptions {
  onRegistry?: (translator: TranslatorOptions) => void
  onDestroy?: (translator: TranslatorOptions) => void
}

export function defineTranslator<T extends IField, K extends ICell>(
  option: TranslatorOptions<T, K>,
) {
  const { roles } = option
  option.roles = [...(roles ?? []), FieldRole.NORMAL]
  return {
    ...option,
  }
}

interface CellTranslatorOptions {
  translators: Array<TranslatorOptions>
}

export class CellTranslator {
  private translators: Record<number, Translator["translate"]> = {}
  private refreshList: Array<Function> = []
  private normalizations: Record<number, Translator["normalization"]> = {}
  private asyncMethods: Record<number, Translator["asyncMethod"]> = {}
  public supportTypes: Array<FieldType> = []
  public caches: Record<string, any> = {}
  public asyncTypes: Array<FieldType> = []
  public autoTypes: Array<FieldType> = []
  public hasOptionsTypes: Array<FieldType> = []
  private roleMap: Record<FieldRole, Array<FieldType>> = {
    [FieldRole.AUTO]: this.autoTypes,
    [FieldRole.HAS_OPTIONS]: this.hasOptionsTypes,
    [FieldRole.ASYNC]: this.asyncTypes,
    [FieldRole.NORMAL]: this.supportTypes,
  }

  constructor(options: CellTranslatorOptions) {
    const { translators } = options
    translators.forEach((translator) => {
      this.registryTranslator(translator)
    })
  }

  private registryTranslator(translator: TranslatorOptions) {
    const { refresh, translate, normalization, cache, asyncMethod, roles } =
      translator
    if (refresh && typeof refresh === "function") {
      this.refreshList.push(refresh)
    }
    if (cache) {
      this.caches[translator.fieldType] = cache
    }
    if (asyncMethod && typeof asyncMethod === "function") {
      this.asyncMethods[translator.fieldType] = asyncMethod
    }
    if (roles) {
      roles.forEach((role) => {
        this.roleMap[role].push(translator.fieldType)
      })
    }
    if (translate && typeof translate === "function") {
      this.translators[translator.fieldType] = translate
    }
    this.normalizations[translator.fieldType] = normalization
  }

  public async getCell(
    field: IField,
    fieldMap: fieldMap,
    value: string | IOpenCellValue,
  ) {
    try {
      if (typeof value !== "string") return field.createCell(value)
      const type = fieldMap.field.type
      const translator = this.translators[type]
      if (!translator) {
        return null
      }
      return await translator(value, field, fieldMap.config)
    } catch (error) {
      Error({
        title: `CellTranslator.getCell: Translate ${
          fieldMap.field.name
        }[FieldType: ${FieldNameList[fieldMap.field.type]}] error`,
        message: String(error),
        error,
      })
      return null
    }
  }

  public refresh() {
    this.refreshList.forEach((refresh) => {
      refresh()
    })
  }

  public async normalization(
    value: string,
    fieldMap: fieldMap,
    config?: fieldMap["config"],
  ) {
    const normalization = this.normalizations[fieldMap.field.type]
    if (!normalization) {
      return value
    }
    return normalization(value, config ?? fieldMap.config)
  }

  public async asyncMethod(
    options: AsyncParams,
    fieldMap: fieldMap,
    config?: fieldMap["config"],
  ) {
    const asyncMethod = this.asyncMethods[fieldMap.field.type]
    if (!asyncMethod) {
      return
    }
    return asyncMethod(options, config ?? fieldMap.config)
  }
}
