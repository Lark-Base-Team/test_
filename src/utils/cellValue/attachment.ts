import {
  type IAttachmentField,
  FieldType,
  bitable,
  UploadFileTaskStatus,
  IOpenAttachment,
} from "@lark-base-open/js-sdk"
import { defineTranslator, type AsyncParams, FieldRole } from "./cell"
import { downLoadFile, unique, validateUrl } from "@/utils"
import type { fieldMap } from "@/types/types"

enum DownloadStatus {
  Waiting = -1,
  Downloading = 0,
  Downloaded = 200,
  Error = 404,
}

interface FileCacheItem {
  url: string
  file: File | null
  name: string
  total: number
  loaded: number
  status: DownloadStatus
  token: string
  timeStamp: number
}

let cache: Record<string, FileCacheItem> = {}

const setCache = (item: FileCacheItem | Array<FileCacheItem>) => {
  if (Array.isArray(item)) {
    item.forEach((i) => {
      cache[i.url] = i
    })
    return
  }
  cache[item.url] = item
}

/**
 * Download files
 *
 * @param urls
 * @param options
 * @returns
 */
function processDownloadFiles(
  urls: Array<string>,
  options?: {
    onProgress?: AsyncParams<string>["onProgress"]
    onError?: AsyncParams<string>["onError"]
    parallel?: number
    requestConfig?: fieldMap["config"]["requestConfig"]
  },
) {
  const cacheItems = unique(urls)
    .map((url) => {
      if (!validateUrl(url)) return
      const item =
        cache[url] ||
        ({
          url,
          file: null,
          name: "",
          total: 0,
          loaded: 0,
          status: DownloadStatus.Waiting,
          token: "",
        } as FileCacheItem)
      return item
    })
    .filter((item) => {
      return (
        item &&
        [DownloadStatus.Waiting, DownloadStatus.Error].includes(item.status)
      )
    }) as Array<FileCacheItem>
  if (!cacheItems.length) return
  setCache(cacheItems)
  const { onProgress, parallel = 4, onError } = options ?? {}
  const total = cacheItems.length
  let loaded = 0
  let runningNum = 0
  const pendingItems: Array<FileCacheItem> = []
  const _run = () => {
    while (runningNum < parallel && cacheItems.length) {
      const item = cacheItems.shift()
      if (!item) return
      item.status = DownloadStatus.Downloading
      runningNum++
      pendingItems.push(item)
      downLoadFile(item.url, {
        onProgress: (p) => {
          console.log("processDownloadFiles", p)
          item.loaded = p.loaded
          item.total = p.total
          onProgress?.({
            total,
            loaded,
            message: pendingItems
              .map(
                (i) =>
                  `Downloading ${i.url} ${((i.loaded / i.total) * 100).toFixed(
                    2,
                  )}%`,
              )
              .join("\n"),
          })
        },
        fetchOptions: options?.requestConfig,
      })
        .then((file) => {
          item.file = file
          item.name = file?.name ?? ""
          item.status = DownloadStatus.Downloaded
          console.log("res", item, file)
        })
        .catch((e) => {
          item.status = DownloadStatus.Error
          onError?.(e)
          console.error(e)
        })
        .finally(() => {
          runningNum--
          loaded++
          console.log("cache", cache)
          pendingItems.splice(pendingItems.indexOf(item), 1)
          console.log("cache", cache)
          _run()
        })
    }
  }
  return new Promise((resolve) => {
    console.log("processDownloadFiles", cacheItems)
    _run()
    const timer = setInterval(() => {
      if (!cacheItems.length) {
        clearInterval(timer)
        resolve(null)
      }
    }, 1000)
  })
}

async function normalization(value: string, config?: fieldMap["config"]) {
  const { separator = "," } = config || {}
  const urls = value.split(separator)
  return urls
}

async function attachment(value: string, field: IAttachmentField) {
  const urls = await normalization(value)
  const files = urls
    .map((url) => {
      const item = cache[url]
      if (!item) return null
      return {
        name: item.name,
        token: item.token,
        size: item.total,
        type: item.file?.type ?? "",
        timeStamp: new Date().getTime(),
      }
    })
    .filter((item) => item) as IOpenAttachment[]
  if (!files) return null
  return await field.createCell(files)
}

export const AttachmentTranslator = defineTranslator({
  fieldType: FieldType.Attachment,
  translate: attachment,
  normalization: async (value: string) => value,
  name: "Attachment",
  asyncMethod: async (
    options: AsyncParams<string>,
    config?: fieldMap["config"],
  ) => {
    const { data: urls, onProgress, onError } = options
    const { requestConfig } = config || {}
    if (!urls.length) return
    await processDownloadFiles(urls, { onProgress, onError, requestConfig })
    const hasFileItems = Object.values(cache).filter((item) => item.file)
    const files = hasFileItems.map((item) => item.file) as File[]
    const off = bitable.base.onUploadStatusChange(({ data }) => {
      const { tasks } = data
      const { list } = tasks
      onProgress?.({
        loaded: list.filter((t) => t.status === UploadFileTaskStatus.Success)
          .length,
        total: list.length,
        message: list
          .filter((t) => t.status === UploadFileTaskStatus.Pending)
          .map(
            (t) =>
              `Uploading ${t.name} ${((t.uploadedSize / t.size) * 100).toFixed(
                2,
              )}`,
          )
          .join("\n"),
      })
    })
    const tokens = await bitable.base.batchUploadFile(files)
    let index = 0
    const items = Object.values(hasFileItems)
    for (const t in items) {
      const item = items[t]
      if (!item.file) continue
      item.token = tokens[index]
      index += 1
    }
    off()
  },
  roles: [FieldRole.ASYNC],
  refresh: () => {
    cache = {}
  },
})
