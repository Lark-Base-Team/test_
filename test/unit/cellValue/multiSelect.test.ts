import { multiSelect } from "../../../src//utils/cellValue/multiSelect"
import { describe, it, expect } from "vitest"
import {
  checkers,
  IMultiSelectFieldMeta,
  FieldType,
  SelectOptionsType,
} from "@lark-base-open/js-sdk"

const field: IMultiSelectFieldMeta = {
  type: FieldType.MultiSelect,
  id: "feldxxxx",
  name: "多选",
  isPrimary: false,
  description: {
    content: [],
  },
  property: {
    options: [
      {
        id: "optxxx1",
        name: "选项1",
        color: 1,
      },
      {
        id: "optxxx2",
        name: "选项2",
        color: 2,
      },
      {
        id: "optxxx3",
        name: "选项3",
        color: 3,
      },
    ],
    optionsType: SelectOptionsType.STATIC,
  },
}
describe("multiSelect", () => {
  it("should return IOpenMultiSelect", () => {
    expect(
      checkers.isMultiSelect(multiSelect("选项1,选项2,选项3", field)),
    ).toBe(true)
    expect(multiSelect("选项1,选项2,选项3", field)).toMatchSnapshot()
  })
})
