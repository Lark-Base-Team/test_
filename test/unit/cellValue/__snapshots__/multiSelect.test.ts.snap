// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`multiSelect > should return IOpenMultiSelect 1`] = `
[
  {
    "id": "optxxx1",
    "text": "选项1",
  },
  {
    "id": "optxxx2",
    "text": "选项2",
  },
  {
    "id": "optxxx3",
    "text": "选项3",
  },
]
`;
